libvast/src/logical_pipeline.cpp

```diff
@@ -0,0 +1,421 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "vast/logical_pipeline.hpp"
+
+#include "vast/concept/parseable/vast/pipeline.hpp"
+#include "vast/element_type.hpp"
+#include "vast/logical_operator.hpp"
+#include "vast/operator_control_plane.hpp"
+#include "vast/physical_operator.hpp"
+#include "vast/plugin.hpp"
+
+#include <queue>
+#include <unordered_map>
+
+namespace vast {
+
+namespace {
+
+class execute_ctrl final : public operator_control_plane {
+public:
+  explicit execute_ctrl() noexcept = default;
+
+  [[nodiscard]] auto error() const noexcept -> const caf::error& {
+    return error_;
+  }
+
+private:
+  [[nodiscard]] auto self() noexcept -> caf::event_based_actor* override {
+    die("not implemented");
+  }
+
+  auto abort(caf::error error) noexcept -> void override {
+    error_ = std::move(error);
+  }
+
+  auto warn(caf::error) noexcept -> void override {
+    die("not implemented");
+  }
+
+  auto emit(table_slice) noexcept -> void override {
+    die("not implemented");
+  }
+
+  auto demand(type = {}) const noexcept -> size_t override {
+    die("not implemented");
+  }
+
+  /// Access available schemas.
+  auto schemas() const noexcept -> const std::vector<type>& override {
+    die("not implemented");
+  }
+
+  /// Access available concepts.
+  auto concepts() const noexcept -> const concepts_map& override {
+    die("not implemented");
+  }
+
+  caf::error error_ = {};
+};
+
+/// Creates a generator from a source operator.
+/// @pre `op->input_element_type()` must be `void`
+auto producer_from_logical_source(runtime_logical_operator* op,
+                                  operator_control_plane* ctrl)
+  -> generator<runtime_batch> {
+  auto gen = op->make_runtime_physical_operator({}, ctrl);
+  if (not gen) {
+    ctrl->abort(gen.error());
+    return {};
+  }
+  auto f = []<element_type Input, element_type Output>(
+             physical_operator<Input, Output> gen) -> generator<runtime_batch> {
+    if constexpr (not std::is_void_v<Input>) {
+      die("tried to instantiate pipeline that does not begin with void");
```

Comment:
If we get here then we have a logic error; in a closed logical pipeline, the `void` element type occurs only at the very start and at the very end. I think terminating on a precondition violation is fine.

Even if we checked outside of the visitor, we would also still need this code inside of it, as the visitor must be exhaustive.

Reply:
That's not what I meant. We handle the fact that output of the previous matches the input of the current operator in `make`, afterwards it is assumed true, but here we need to check for it again nonetheless. So it would be reasonable to add a pointer to the validation code.

Reply:
Like this?

```suggestion
    if constexpr (not std::is_void_v<Input>) {
      // Only the first operator can begin with the element type void; this
      // pre-condition is enforced and checked in the `logical_pipeline::make`
      // function.
      die("tried to instantiate pipeline that does not begin with void");
```

Reply:
I have added an explanation to every call to `die`. (The check referred to in the suggestion above is located in `make_local_executor`, as `logical_pipeline::make` allows open pipelines).